lib/async_hooks.js

```diff
@@ -461,6 +462,18 @@ function init(asyncId, type, triggerId, resource) {
   processing_hook = false;
 }
 
+const JSProviders = new Set();
+function registerTypeName(type) {
+  if (JSProviders.has(type))
+    throw new errors.TypeError('ERR_ASYNC_PROVIDER_NAME', type);
+  JSProviders.add(type);
+  return type;
+}
+
+function getTypes() {
+  return Object.keys(async_wrap.Providers)
+    .concat(...JSProviders.keys());
```

Comment:
No need for `.keys()`. `Set`'s iteration protocol returns its contents by default.